### PR TITLE
IGNITE-10983 Data region validation on joining nodes.

### PR DESCRIPTION
**Pull request by ingvard**

### CODE REVIEW

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/CacheWithDifferentDataRegionConfigurationTest.java [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+    private static final String CACHE_2 = "cache_2";
+
+    /** Persistence. */
+    private static final boolean PERSISTENCE = true;
+
+    /** Memory. */
+    private static final boolean MEMORY = false;
+
+    /**
+     * @throws Exception
```
Comment (glukos) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/CacheWithDifferentDataRegionConfigurationTest.java]: Please add 'throws' tag description

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/CacheWithDifferentDataRegionConfigurationTest.java [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+     */
+    @After
+    public void tearDown() throws Exception {
+        stopAllGrids();
+        cleanPersistenceDir();
+    }
+
+    /**
+     * @throws Exception
```
Comment (glukos) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/CacheWithDifferentDataRegionConfigurationTest.java]: Please add 'throws' tag description

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/CacheWithDifferentDataRegionConfigurationTest.java [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+            .withRegion(REGION_1, PERSISTENCE)
+            .start();
+
+        assertThrowsContainsMessage(() -> node(NODE_2).withRegion(REGION_1, MEMORY).start(),
+            IgniteSpiException.class,
+            "Failed to join node (Incompatible data region configuration [region=" + REGION_1
+        );
+    }
+
+    @Test
```
Comment (glukos) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/CacheWithDifferentDataRegionConfigurationTest.java]: Please add javadoc (at least empty)

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/CacheWithDifferentDataRegionConfigurationTest.java [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+    public void secondNodeMustRejectJoinOnThirdNode() throws Exception {
+        IgniteEx node1 = node(NODE_1)
+            .start();
+
+        node1.cluster().baselineAutoAdjustTimeout(1); //Hack: The way to add persistence cache into in-memory cluster
+
+        IgniteEx node2 = node(NODE_2)
```
Comment (glukos) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/CacheWithDifferentDataRegionConfigurationTest.java]: variable is never used

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/CacheWithDifferentDataRegionConfigurationTest.java [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+        }
+
+        /** Start node from builder */
+        public IgniteEx start() throws Exception {
+            IgniteConfiguration cfg = getConfiguration(gridName);
+
+            cfg.setConsistentId(gridName);
+
+            DataStorageConfiguration storageCfg = new DataStorageConfiguration();
+            storageCfg.setDataRegionConfigurations(regions.toArray(new DataRegionConfiguration[] {}));
```
Comment (glukos) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/CacheWithDifferentDataRegionConfigurationTest.java]: looks suspicious. maybe this way? v
```suggestion
            storageCfg.setDataRegionConfigurations(regions.toArray(newDataRegionConfiguration[regions.size()]));

```

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java [hunk 6]
```diff
@@ -1059,6 +1068,32 @@ private void initializeInternalCacheNames() {
             internalCaches.add(CU.SYS_CACHE_HADOOP_MR);
     }
 
+    /**
+     * @param node
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java]: Please add tag description

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheUtils.java [hunk 7]
```diff
@@ -1888,6 +1890,21 @@ private static DataStorageConfiguration extractDataStorage(ClusterNode node, Ign
         return null;
     }
 
+    /**
+     * @return {@code true} if persistence is enabled for a default data region, {@code false} if not.
+     */
+    public static boolean isDefaultDataRegionPersistence(DataStorageConfiguration cfg) {
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheUtils.java]: sounds strange.
maybe 'isDefaultDataRegionPersistent'?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheUtils.java [hunk 8]
```diff
@@ -1888,6 +1890,21 @@ private static DataStorageConfiguration extractDataStorage(ClusterNode node, Ign
...
+        if (cfg == null)
+            return false;
+
+        DataRegionConfiguration defaultRegionCfg = cfg.getDefaultDataRegionConfiguration();
```
Comment (glukos) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheUtils.java]: abbreviation plugin suggests to change name to 'dfltRegionCfg'

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java (hunk 6 above)
Comment (ingvard) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java]: Added

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheUtils.java (hunk 7 above)
Comment (ingvard) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheUtils.java]: It sounds better. Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheUtils.java (hunk 8 above)
Comment (ingvard) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheUtils.java]: Fixed

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/CacheWithDifferentDataRegionConfigurationTest.java (hunk 3 above)
Comment (ingvard) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/CacheWithDifferentDataRegionConfigurationTest.java]: Fixed

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/CacheWithDifferentDataRegionConfigurationTest.java (hunk 4 above)
Comment (ingvard) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/CacheWithDifferentDataRegionConfigurationTest.java]: Fixed

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/CacheWithDifferentDataRegionConfigurationTest.java (hunk 5 above)
Comment (ingvard) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/CacheWithDifferentDataRegionConfigurationTest.java]: if (a.length < size) return (T[]) Arrays.copyOf(elementData, size, a.getClass());

It creates an array inside. Fixed it.

Comment (ingvard) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/CacheWithDifferentDataRegionConfigurationTest.java]: For example, the IDEA suggests: .toArray(new DataRegionConfiguration[0])

**Pull request closed without merging**